### DISCUSSION TITLE
SSL certificate verification error while sending a notification to slack channel

### DISCUSSION THREAD

**Discussion by Kannedhara** [apache/airflow#21854] [General]
### Apache Airflow Provider(s)

slack

### Versions of Apache Airflow Providers

4.0.2

### Apache Airflow version

2.2.2

### Operating System

Amazon Managed Airflow

### Deployment

MWAA

### What happened

I am using a DAG which has sending slack notifications as TASK. I am using SlackWebHookOperator to send notifications. I create a web hook and created a connection. I am using the same connection in SlackWebHookOperator. I am getting "ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate in certificate chain (_ssl.c:1091)" error. I see, slack made a mandate SSL verification from march 4,2020.

Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/urllib3/connectionpool.py", line 706, in urlopen
    chunked=chunked,
  File "/usr/local/lib/python3.7/site-packages/urllib3/connectionpool.py", line 382, in _make_request
    self._validate_conn(conn)
  File "/usr/local/lib/python3.7/site-packages/urllib3/connectionpool.py", line 1010, in _validate_conn
    conn.connect()
  File "/usr/local/lib/python3.7/site-packages/urllib3/connection.py", line 426, in connect
    tls_in_tls=tls_in_tls,
  File "/usr/local/lib/python3.7/site-packages/urllib3/util/ssl_.py", line 450, in ssl_wrap_socket
    sock, context, tls_in_tls, server_hostname=server_hostname
  File "/usr/local/lib/python3.7/site-packages/urllib3/util/ssl_.py", line 493, in _ssl_wrap_socket_impl
    return ssl_context.wrap_socket(sock, server_hostname=server_hostname)
  File "/usr/lib64/python3.7/ssl.py", line 423, in wrap_socket
    session=session
  File "/usr/lib64/python3.7/ssl.py", line 870, in _create
    self.do_handshake()
  File "/usr/lib64/python3.7/ssl.py", line 1139, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate in certificate chain (_ssl.c:1091)

### What you expected to happen

I used SlackWebHookOperator in airflow to send notifications in 2019 and I was not facing any issues earlier. Due to above error, we are unable to send notifications to slack using slack web hook

### How to reproduce

Create a slack channel, create a incoming webhook for the channel. Using webhook URL, create a slack connection under connections section. Try to send notifications to slack channel using SlackWebHookOperator

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by potiuk** [comment]
Converted it into discussion. I think you simply should make a. non-self-signed certificate for your webhook. The error is - I believe - not raised by Airlfow but by Slack trying to connect to your server (but I am guessing as you only provided partial stacktrace and it's not clear where it comes from).

My recommendation is to use https://letsencrypt.org/ to get the proper non-self-signed certificate so that Slack can verify it.

**Comment by irudayambics** [comment]
I encountered the same issue today with airflow 2.2.3. I tried adding custom CA to ubuntu trust store and adding REQUESTS_CA_BUNDLE=/path/to/your/certificate.pem environment variable but still the error 'ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate in certificate chain (_ssl.c:1131)' 

apache-airflow                  2.2.3
apache-airflow-providers-celery 2.1.0
apache-airflow-providers-ftp    2.0.1
apache-airflow-providers-http   2.0.1
apache-airflow-providers-imap   2.0.1
apache-airflow-providers-sqlite 2.0.1

Flask                           1.1.2
Flask-AppBuilder                3.4.1
Flask-Babel                     2.0.0
Flask-Caching                   1.10.1
Flask-JWT-Extended              3.25.1
Flask-Login                     0.4.1
Flask-OpenID                    1.3.0
Flask-SQLAlchemy                2.5.1
Flask-WTF                       0.14.3

**Reply by bh-ashishsharma** [reply]
Is there a workaround around it? I have facing a similar issue when using a self-signed certificate?

**Reply by potiuk** [reply]
Yep. change it to non-self-signed certificate (I suggest letsencrypt). 

This is not Airflow problem again. this is choice of library and slack developers. Bring it there if you think it is wrong.

**Reply by irudayambics** [reply]
@bh-ashishsharma I resolved by following the steps below because I am using Custom CA.

Step 1: add REQUESTS_CA_BUNDLE=/path/to/your/certificate.pem to /etc/environment file 
Step 2: Add /etc/environment to airflow-webserver service file
Step 3: Restart airflow-webserver

You may to try exporting your Root CA using the link below and add it to your environment file as well.

https://support.kaspersky.com/CyberTrace/1.0/en-US/174127.htm

**Comment by miahstewart** [comment]
I get a similar error **only** when using the new @task decorator.

> (Caused by SSLError(SSLError("bad handshake: Error([('SSL routines', 'tls_process_server_certificate', 'certificate verify failed')])")))
[2022-03-15 20:59:52,292] {{local_task_job.py:264}} INFO - 0 downstream tasks scheduled from follow-on schedule check


We have legacy DAGs which are running fine, but the one DAG using the decorator silent errors (no error is sent to slack when calling the on_failure_callback).

Is there some conflict going on here?
In my airflow_slack_failed_task_webhook, the failing line is the urllib3 line as follows:
`urllib3.PoolManager().request('POST', 'https://hooks.slack.com/services/' + token
   , body=slack_msg, headers={'Content-Type': 'application/json'})`

Do I need to manually specify the proxy parameters? I have tried manually specifying the certifi certificate location and updating that to the latest available version.


Example below:

Silent error:

`@task(task_id=f'run_schedule_plan_{item_name}', on_failure_callback=airflow_slack_failed_task_webhook)
            def run_scheduled_plan(schedule_item):
                dashboard_id = schedule_item.get('dashboard_id')
                item_name = schedule_item.get('name')
                
                sdk = looker_login()
                response = run_new_scheduled_plan_and_check(sdk, dashboard_id,
                                        scheduled_plan_name=f'{schedule_group.replace("_", " ") + (" - " + item_name if item_name else "")}'.title(),
                                        delivery = delivery, slack_channel = slack_channel, emails = emails)`

and what runs fine as usual:

`    t_update_tables = PythonOperator(task_id             = 'update_tables',
                                           python_callable     = update_tables,
                                           op_kwargs           = {'tables'         : tables,
                                                                  'our_database' : our_database,
                                                                  'etl_database'   : etl_database,
                                                                  'warehouse'      : warehouse},
                                           provide_context     = True,
                                           dag                 = dag,
                                           on_failure_callback = airflow_slack_failed_task_webhook)`

**Reply by miahstewart** [reply]
If anyone is interested, this is actually a problem isolated to the looker_sdk. There appears to be a conflict between urllib3 and the python looker_sdk.

**Comment by Kannedhara** [comment]
We have tried different options for SSL but nothing seems worked. We are continuing to work on different options. I don't have a way to install self certificate as it is a managed service from AWS.

As a work around for this problem, I used the below code to suppress SSL verification and it worked.
requests.post('https://slack.com/api/chat.postMessage', {
            'token': slack_token,
            'channel': slack_channel,
            'text': text,
            'icon_emoji': slack_icon_emoji,
            'username': "gopi",
            'blocks': json.dumps(blocks) if blocks else None
        },verify=False).json()


@potiuk  I am trying to suppress the SSL using SlackWebHookOperator but it is not working.  I tried the below code.

SlackWebhookOperator(
            task_id='slack_webhook',
            http_conn_id=SLACK_CONN_ID,
            webhook_token=SLACK_WEBHOOK_TOKEN,
            message=slack_msg,
            username='airflow',
            extra_options={'verify': False}
        )

Any thoughts ?

**Comment by miahstewart** [comment]
@Kannedhara 
I managed to fix this by changing what we currently had (which SSL errored):

`
response = (urllib3.PoolManager()
                .request('POST', 'https://hooks.slack.com/services/' + token, body=slack_msg, headers={'Content-Type': 'application/json'}))`

to (which now works)

`
response = (urllib3.PoolManager(cert_reqs='CERT_REQUIRED', ca_certs=certifi.where()) 
                .request('POST', 'https://hooks.slack.com/services/' + token, body=slack_msg, headers={'Content-Type': 'application/json'}))`

I'm not sure why, but I think at a certain point in time the `requests` package demanded this explicit certification stuff.
